server/routes/companies.js

```diff
@@ -116,25 +179,399 @@ router.get('/filter', function(req, res, next) {
   const mid = req.query.mid;
 
   if (pid && mid) {
-    db.query('SELECT DISTINCT company_major.company_id FROM company_major LEFT OUTER JOIN company_position ON (company_major.company_id = company_position.company_id) WHERE (position_id = $1 AND major_id = $2);', [pid, mid], (err, result) => {
-      if (err) return next(err);
-      res.send(result.rows);
-    });
+    knex.distinct('company_major.company_id').from('company_major')
+    .leftOuterJoin('company_position', 'company_major.company_id', 'company_position.company_id')
+    .where({position_id: req.body.pid, major_id: req.query.mid})
+      .then(result => {
+        if (result.length) {
+          res.json(result);
+        } else {
+          util.throwError(404, 'No companies found');
+        }
+      })
+      .catch(err => { return next(err) });
   } else if (mid) {
-    db.query('SELECT company_id FROM company_major WHERE major_id = $1;', [mid], (err, result) => {
-      if (err) return next(err);
-      res.send(result.rows);
-    });
+    knex('company_major').select('company_id').where({major_id: req.body.mid})
+      .then(result => {
+        if (result.length) {
+          res.json(result);
+        } else {
+          util.throwError(404, 'No companies found');
+        }
+      })
+      .catch(err => { return next(err) });
   } else if (pid) {
-    db.query('SELECT company_id FROM company_position WHERE position_id = $1;', [pid], (err, result) => {
-      if (err) return next(err);
-      res.send(result.rows);
-    });
+    knex('company_position').select('company_id').where({position_id: req.body.pid})
+      .then(result => {
+        if (result.length) {
+          res.json(result);
+        } else {
+          util.throwError(404, 'No companies found');
+        }
+      })
+      .catch(err => { return next(err) });
   } else { 
-    return res.status(400).send({
-      message: '\'mid\' and \'pid\' are undefined'
-    });
+    util.throwError(400, '\'mid\' and \'pid\' are undefined');
   }
 });
 
+// Add a single company or related data
+router.post('/', function(req, res, next) {
+  values = {
+    name: req.body.name,
+    website: req.body.website,
+    logo: req.body.logo,
+    citizenship_requirement: req.body.citizenship_requirement,
+    description: req.body.description
+  }
+  let ranks = req.body.rank;
+  let user_ids = req.body.user_id;
+  let position_ids = req.body.position_id;
+  let major_ids = req.body.major_id;
+  let contact_ids = req.body.contact_id;
+  let event_ids = req.body.event_id;
+
+  console.log(req.body.position_id);
+
+  var queryCompany = knex('company').insert(values)
+  knex.transaction(function(trx) {
+    return queryCompany.transacting(trx)
+      .returning('id')
+      .then(async function(ids){
+        let company_id = ids[0];
+
+        if (ranks && user_ids) {
+          let rankValues = [];
+          if(Array.isArray(ranks) && Array.isArray(user_ids)){
```

Comment:
You'll have to check, but I think there are simpler ways of seeing whether a param returns an array or not. Try using just .length?

Reply:
We tried this, but we ran into a weird error where only one param with multiple digits would be treated as an array (which each element being a digit) rather than a single entry. This method resolved the issue

Reply:
What happens if you pass ranks[] as the key in postman? does it default to array even with a single value? I've seen people pass it with this syntax before, maybe for this purpose

Reply:
Ok actually just stick with Array.isArray then